src/Framework/LazyFormattedBuildEventArgs.cs

```diff
@@ -17,33 +16,34 @@ public class LazyFormattedBuildEventArgs : BuildEventArgs
         /// <summary>
         /// Stores the message arguments.
         /// </summary>
-        private object[] arguments;
+        private volatile object argumentsOrFormattedMessage;
 
         /// <summary>
         /// Exposes the underlying arguments field to serializers.
         /// </summary>
         internal object[] RawArguments
         {
-            get => arguments;
-            set => arguments = value;
+            get => (argumentsOrFormattedMessage is object[] arguments) ? arguments : null;
         }
 
         /// <summary>
-        /// Stores the original culture for String.Format.
+        /// Exposes the formatted message string to serializers.
         /// </summary>
-        private string originalCultureName;
+        private protected override string FormattedMessage
+        {
+            get => (argumentsOrFormattedMessage is string formattedMessage) ? formattedMessage : base.FormattedMessage;
+        }
 
         /// <summary>
-        /// Non-serializable CultureInfo object
+        /// Stores the original culture for String.Format.
         /// </summary>
-        [NonSerialized]
-        private CultureInfo originalCultureInfo;
+        private string originalCultureName;
```

Comment:
Let's think if we need to store the culture name AND the actual culture object. How are they used? Can we reduce two fields to just one, perhaps?